crates/oci-distribution/src/regexp.rs

```diff
@@ -0,0 +1,16 @@
+use regex::{Regex, RegexBuilder};
+
+/// REFERENCE_REGEXP is the full supported format of a reference. The regexp
+// is anchored and has capturing groups for name, tag, and digest components.
+pub const REFERENCE_REGEXP: &str = r"^((?:(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(?:(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+)?(?::[0-9]+)?/)?[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?::([\w][\w.-]{0,127}))?(?:@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$";
```

Comment:
I assume this is the regular expression used by upstream tools (`containerd`?).

Could you link the original file where this is defined?

Reply:
Sure!

https://github.com/docker/distribution/blob/2800ab02245e2eafc10e338939511dd1aeb5e135/reference/regexp.go#L58-L70